### ISSUE TITLE
Hard faults when inserting a billion records into HTreeMap

### ISSUE THREAD

**Issue by jlnicolson** [jankotek/mapdb#363]
Hi,

I'm trying to store about 1 billion records using mapdb as a (non-in-memory) persistence store. Unfortunately, it seems as though mapdb is only capable of storing around 150 million records. I was wondering whether anyone could verify this. I'm running mapdb on a 4-core 2.66GHz machine with 16GB RAM, the JVM has the parameters -Xms256m -Xmx15G. My input file is simply a list of big integers stored as hex. My code is as follows:

```
package test.mapdb;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

import org.mapdb.Bind;
import org.mapdb.DB;
import org.mapdb.DBMaker;
import org.mapdb.Fun;
import org.mapdb.HTreeMap;

import java.util.NavigableSet;

public class App 
{
	private static int i = 0;
	

	private static String BILLION_BIGINTS_PATH = ".\billion-bigints.txt";
	private static String CURRENT_BIGINTS_PATH = BILLION_BIGINTS_PATH;
	private static long CURRENT_BIGINTS_COUNT = 1000000000;
	
	private static String MAPDB_FILENAME = "map-billion.db";
	
	private static SimpleDateFormat ft = new SimpleDateFormat ("E yyyy.MM.dd 'at' hh:mm:ss a zzz");
		
	public static void createMapDb() throws IOException
	{
		DB db = DBMaker.newFileDB(new File(MAPDB_FILENAME))
				.mmapFileEnablePartial()
				.closeOnJvmShutdown()
				.cacheDisable()
				.transactionDisable()
				.asyncWriteEnable()
				.make();
		
		
        HTreeMap<Integer,BigInteger> map = db.getHashMap("map");

        // inverse mapping for primary map
        //NavigableSet<Fun.Tuple2<String, Long>> inverseMapping = new TreeSet<Fun.Tuple2<String, Long>>();
        //NOTE: you may also use Set provided by MapDB to make it persistent
        //Set<Fun.Tuple2<BigInteger, Integer>> inverseMapping = db.getHashSet("set");
        HTreeMap<BigInteger,Integer> rMap = db.getHashMap("reverseMap");
        
        // bind inverse mapping to primary map, so it is auto-updated
        Bind.mapInverse(map, rMap);
		
        BufferedReader br = new BufferedReader(new FileReader(CURRENT_BIGINTS_PATH));
        long line_count = CURRENT_BIGINTS_COUNT;
        int ten_percent = (int) line_count / 1000;
        
        long start = System.currentTimeMillis();
        String line;
        while ((line = br.readLine()) != null)
        {
        	if(i % ten_percent == 0)
        	{
        		Date now = new Date();
        		System.out.println("[" + ft.format(now) + "] i: " + i);
        	}
        	BigInteger b = new BigInteger(line, 16);
        	map.put(i, b);
        	i++;
        }
    	long stop = System.currentTimeMillis();
    	long diff = stop - start;
    	double peritem = diff / (i * 1.0); 
    	System.out.println("n: " + i + " total time in ms: " + diff + " per-item: " + peritem);
        
    	db.commit();
    	db.close();
	}
	
	public static void getMapDb() throws IOException
	{

		Date now = new Date();
		System.out.println("[" + ft.format(now) + "] reading database");
		
		DB db = DBMaker.newFileDB(new File(MAPDB_FILENAME))
				.mmapFileEnable()
				.cacheDisable()
				.transactionDisable()
				.asyncWriteEnable()
				.make();
		
		now = new Date();
		//System.out.println("[" + ft.format(now) + "] compacting database");
		//db.compact();
		
        HTreeMap<Integer,BigInteger> map = db.getHashMap("map");

        // inverse mapping for primary map
        //NavigableSet<Fun.Tuple2<String, Long>> inverseMapping = new TreeSet<Fun.Tuple2<String, Long>>();
        //NOTE: you may also use Set provided by MapDB to make it persistent
        //Set<Fun.Tuple2<BigInteger, Integer>> inverseMapping = db.getHashSet("set");
        HTreeMap<BigInteger,Integer> rMap = db.getHashMap("reverseMap");
        
        
        // bind inverse mapping to primary map, so it is auto-updated
        Bind.mapInverse(map, rMap);
		
        long start = System.currentTimeMillis();


        BufferedReader br = new BufferedReader(new FileReader(CURRENT_BIGINTS_PATH));
        long line_count = CURRENT_BIGINTS_COUNT;
        int ten_percent = (int) line_count / 10;
        for(i = 0 ; i <line_count; i++)
        {
        	if(i % ten_percent == 0)
        	{
        		long diff = System.currentTimeMillis() - start;
        		now = new Date();
				System.out.println("[" + ft.format(now) + "] ms: " + diff + " i: " + i);
        	}
        	BigInteger b = map.get(i);
        	if (b == null)
        		System.err.println("b is null at i = " + i );
        }
    	long stop = System.currentTimeMillis();
    	long diff = stop - start;
    	double peritem = diff / (i * 1.0); 
    	System.out.println("n: " + i + " total time in ms: " + diff + " per-item: " + peritem);

    	i = 0;
        start = System.currentTimeMillis();
        String line;
        while((line = br.readLine()) != null)
        {
        	if(i % ten_percent == 0)
        	{
        		diff = System.currentTimeMillis() - start;
        		now = new Date();
				System.out.println("[" + ft.format(now) + "] ms: " + diff + " i: " + i);
        	}
        	BigInteger b = new BigInteger(line, 16);
        	
        	Integer out_i = rMap.get(b);
        	
        	if (out_i == null)
        		System.err.println("out_i is null at i = " + i );
        	i++;
        }
    	stop = System.currentTimeMillis();
    	diff = stop - start;
    	peritem = diff / (i * 1.0); 
    	System.out.println("n: " + i + " total time in ms: " + diff + " per-item: " + peritem);
     
    	
    	//db.commit();
    	db.close();
    	throw new IOException();
	}
	
    public static void main( String[] args ) throws IOException
    {
    	try {
    		createMapDb() ;
    		//getMapDb();
		} catch (IOException e) {
			System.out.println("i was: " + i);
			throw e;
		}
    }
}

```

**Comment by jankotek** [comment]
I tested this issue, no problem beyond usual performance. Async Write could cause OutOfMemoryException if not committed regularly, that is solved in 2.0. Also data pump works with HashMap and could be used to make it more efficient.

**Comment by allanwax** [comment]
This may be an old issue at this point, but I also tried to store a billion records in a set.  While technically the failure was not caused by MapDB it was caused by trying to use a memory mapped DB and I simply ran out of memory and swap space to hold it all.

In a sense, there is no magic here, if you memory map a large quantity of memory the OS has to put it somewhere.  I switched to FileChannel and it worked fine.
